### ISSUE TITLE
Query rule serialization is going into infinite loop

### ISSUE THREAD

**Issue by GregoryOrciuch** [algolia/algoliasearch-client-java#492]
- Algolia Client Version: 2.17.3
- Language Version: 8

### Description
Want to use API to save the query rules but the Client refuses me because com.algolia.search.exceptions.AlgoliaException: can not serialize request body ,
that is caused by com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion
from jacksons ObjectMapper, which dont want to serialize the Rule object.

### Steps To Reproduce
If you run this junit test it will throw serialization exception.
Mind that in debugger the Rule object looks pretty well.
I'm trying to reflect what I configured in Algolia UI as this (which is valid and saved via UI rule)
```json
{
  "condition": {
    "pattern": "{facet:products.properties.fbrand}",
    "anchoring": "contains"
  },
  "consequence": {
    "params": {
      "automaticOptionalFacetFilters": [
        "products.properties.fbrand"
      ],
      "query": {
        "remove": [
          "{facet:products.properties.fbrand}"
        ]
      }
    }
  },
  "description": "Boost Brands",
  "objectID": "1528810686725"
}
```

the JUNIT:
```java
import java.util.Collections;

import com.algolia.search.inputs.query_rules.Condition;
import com.algolia.search.inputs.query_rules.Consequence;
import com.algolia.search.inputs.query_rules.ConsequenceParams;
import com.algolia.search.inputs.query_rules.ConsequenceQueryObject;
import com.algolia.search.inputs.query_rules.Rule;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.Test;

public class AlgoliaTest {

	@Test
	public void serializationTest() throws  Exception{

		ObjectMapper objectMapper = new ObjectMapper();

		Condition condition = new Condition()
				.setPattern("{facet:products.properties.fbrand}")
				.setAnchoring("contains");

		ConsequenceQueryObject consequenceQuery = new ConsequenceQueryObject()
				.setRemove(Collections.singletonList("{facet:products.properties.fbrand}"));

		ConsequenceParams params = new ConsequenceParams()
				.setAutomaticFacetFilters(Collections.singletonList("products.properties.fbrand"));

		params.setQuery(consequenceQuery);
		Consequence consequence = new Consequence()
				.setParams(params);

		Rule rule1 = new Rule()
				.setObjectID("1528811588947")
				.setDescription("Boost Brands")
				.setCondition(condition)
				.setConsequence(consequence);

		String result = objectMapper.writeValueAsString(rule1);

	}
}

```

And the result will be:
com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion (StackOverflowError) (through reference chain: com.algolia.search.inputs.query_rules.Rule["consequence"]->com.algolia.search.inputs.query_rules.Consequence["params"]->com.algolia.search.inputs.query_rules.ConsequenceParams["query"])

Thanks for looking!

**Comment by aseure** [comment]
Thank you for submitting your issue. This definitely looks like a bug in our implementation.

Thank you as well for providing us the JUnit test to be able to reproduce. We'll take care of adding it to the test suite shortly and to fix the bug at the same time.

**Comment by GregoryOrciuch** [comment]
Can you share the ETA of a release which will contain a fix for that ?
Thanks!

**Comment by aseure** [comment]
Hello @GregoryOrciuch 

At the moment, I cannot give you a precise ETA but I've already started worked on it as we speak.

**Comment by GregoryOrciuch** [comment]
I will help a little bit.
If you remove those custom serialize/deserializer from there: com.algolia.search.inputs.query_rules.ConsequenceQuery

the test will pass and the output json will look very nice.

just replace your class in there: algoliasearch-client-java-2/algoliasearch-common/src/main/java/com/algolia/search/inputs/query_rules/ConsequenceQuery.java 
with that content:

```java
package com.algolia.search.inputs.query_rules;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.io.Serializable;

@JsonInclude(JsonInclude.Include.NON_NULL)
public abstract class ConsequenceQuery implements Serializable {}

```

but maybe it breaks behavior over com.algolia.search.inputs.query_rules.ConsequenceQueryString object.

**Comment by aseure** [comment]
@ElPicador I've added the test and I was about to add a custom serializer for the `ConsequenceQueryObject` to avoid the infinite recursion but @GregoryOrciuch is actually right: simply removing the custom (de)serializers of `ConsequencQuery` abstract class in the first place is working (all tests + the new unit test are passing).

Any reason I'm not aware of to keep the custom serializers?

**Comment by aseure** [comment]
Fix deployed and available as of [2.18.0](https://github.com/algolia/algoliasearch-client-java-2/blob/master/ChangeLog.md#2180-2018-07-26).

Thank you again for your contribution @GregoryOrciuch.
